### PR TITLE
Adds support for name/value embedded Bindings

### LINKED ISSUE #617: More concise embedded TriggerBindings
# Expected Behavior
A trigger binding consists of name and value pairs. So, when we embed it we should be able to use something like:

```
 bindings:
    - ref: my-cool-binding # this one is a ref to an existing binding

    - name: commit_id # embedded binding
      value: "$(body.head_commit_id)

    - name: message # embedded binding with static values
      value: "Hello from an embedded binding"
```

# Actual Behavior
Today, the above example looks like:
```
 bindings:
        - ref: my-cool-binding # this one is a ref to an existing binding
       
        -  name: some-name
           spec:
              params:
                - name: commit_id # embedded binding
                  value: "$(body.head_commit_id)
                - name: message # embedded binding with static values
                  value: "Hello from an embedded binding"
```

# Additional Info

From: https://github.com/tektoncd/triggers/pull/548/files#r417024864
/kind feature

### PR DESCRIPTION
**Pull request by dibyom**
# Changes

This commit adds support for a more concise syntax for TriggerBindings embedded
in a Trigger or EventListener:

NEW SYNTAX:
```
bindings:
- ref: some-name
- name: commit_id # embedded binding
  value: "$(body.head_commit_id)
```

OLD SYNTAX
```
bindings:
- name: some-name
  spec:
     params:
     - name: commit_id # embedded binding
       value: "$(body.head_commit_id)
```

The old embedded syntax is deprecated and Triggers will convert any old style
embedded bindings to the new syntax via the MutatingAdmissionWebhook.

BREAKING CHANGE: With this change, users cannot specify both `name` and `ref`
for a single binding. Use `ref` to refer to a TriggerBinding resource and
`name` for embedded bindings.

Fixes #617 and implements [TEP-0016](https://github.com/tektoncd/community/blob/master/teps/0016-concise-trigger-bindings.md)

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/triggers/blob/master/CONTRIBUTING.md) for more details._

# Release Notes
```release-note
action required: TriggerBindings can now be embedded by using just name/value fields inside a Trigger or a EventListener.

BREAKING CHANGE: With this change, users cannot specify both `name` and `ref`
for a single binding. Use `ref` to refer to a TriggerBinding resource and
`name` for embedded bindings.

NEW SYNTAX:
bindings:
- ref: some-name
- name: commit_id # embedded binding
  value: "$(body.head_commit_id)

OLD SYNTAX:
bindings:
- name: some-name
  spec:
    params:
    - name: commit_id # embedded binding
      value: "$(body.head_commit_id)
```

### CODE REVIEW

**Comment by dibyom** [comment]
/hold 
Gonna look into why coverage dropped.

**Comment by dibyom** [comment]
The testbuilder `EventListenerTriggerBinding` allows users to specify both name and ref. With this change, this is not allowed. This was fine for all the happy path tests but tests that are verifying error paths (such as validation errors) are now all failing due to the structs created by builders having both name and refs (this is causing the drop in coverage). To fix, first I'll change the builder to only accept name and ref. Later, we can migrate our tests to not use builders at all and have our error tests verify the actual error string.

File: pkg/apis/triggers/v1alpha1/trigger_validation.go
Comment (wlynch) [inline pkg/apis/triggers/v1alpha1/trigger_validation.go:84]: Let's link to an issue and add comments about when this is safe to delete.

File: pkg/template/resource.go
Comment (wlynch) [inline pkg/template/resource.go:65]: Doesn't look like this is being used outside of this package. Let's unexport this for now until we have a clear need, particularly since this relies on unexported types as parameters.

File: pkg/template/resource.go
Comment (wlynch) [inline pkg/template/resource.go:56]: I'm 50/50 on this - should we remove this? This should already be validated to ensure the name is there (https://github.com/tektoncd/triggers/blob/f73c5f12a03f063e448d414f96e1b18e2ed2185e/pkg/apis/triggers/v1alpha1/trigger_validation.go#L66), and if it is empty it should cause the getTT to fail

File: docs/triggers.md
Comment (wlynch) [inline docs/triggers.md:28]: Let's also update `docs/triggerbindings.md`.

**Review by wlynch — COMMENTED** [review]
/lgtm

Few small comments.

File: docs/triggers.md
Comment (dibyom) [reply docs/triggers.md:28]: `docs/triggerbindings.md` right now contains sections just on the TriggerBinding type and its syntax. The sections on embedded bindings are in the docs/triggers.md and docs/eventlisteners.md

File: pkg/apis/triggers/v1alpha1/trigger_validation.go
Comment (dibyom) [reply pkg/apis/triggers/v1alpha1/trigger_validation.go]: done

File: pkg/template/resource.go
Comment (dibyom) [reply pkg/template/resource.go]: done

File: pkg/template/resource.go
Comment (dibyom) [reply pkg/template/resource.go]: yeah that makes sense. removed.

**Comment by dibyom** [comment]
Ready for another look @wlynch @khrm

File: pkg/template/resource.go
Comment (khrm) [inline pkg/template/resource.go:97]: This looks to be a new behaviour. Should we document this as breaking change?

File: pkg/template/resource.go
Comment (khrm) [inline pkg/template/resource.go:66]: Won't it cause duplication error (https://github.com/tektoncd/triggers/pull/768/files#diff-7b6fd7ee11e66c6319df083beac6f30aR97)  if SetDefaults has already been called?

File: pkg/template/resource.go
Comment (khrm) [inline pkg/template/resource.go:67]: Won't it cause duplication error (https://github.com/tektoncd/triggers/pull/768/files#diff-7b6fd7ee11e66c6319df083beac6f30aR97) if SetDefaults has already been called?

File: pkg/template/resource.go
Comment (dibyom) [reply pkg/template/resource.go:97]: we already had this check...just in a different function before: https://github.com/tektoncd/triggers/blob/v0.8.1/pkg/template/resource.go#L151

File: pkg/template/resource.go
Comment (dibyom) [reply pkg/template/resource.go:67]: ahh good point.

Comment (dibyom) [reply pkg/template/resource.go:67]: actually, looking through the code, I'm not sure it will. In SetDefaults, we convert the Spec.Params to name/value pairs. We do not keep the Spec field after.

The case here will only apply for cases where set defaults has not been called i.e. if someone upgrades to this version with an existing EL that contains an embedded binding.

File: pkg/apis/triggers/v1alpha1/trigger_validation.go
Comment (savitaashture) [inline pkg/apis/triggers/v1alpha1/trigger_validation.go:84]:
```suggestion
		case b.Spec != nil: // TODO(#768): Remove Deprecated old style embedded bindings
```

File: pkg/apis/triggers/v1alpha1/trigger_validation.go
Comment (savitaashture) [inline pkg/apis/triggers/v1alpha1/trigger_validation.go:80]: I can able to specify binding like below

```
apiVersion: triggers.tekton.dev/v1alpha1
kind: EventListener
metadata:
  name: github-listener-interceptor
spec:
  triggers:
    - name: github-listener
      interceptors:
        - github:
            secretRef:
              secretName: github-secret
              secretKey: secretToken
            eventTypes:
              - pull_request
        - cel:
            filter: "body.action in ['opened', 'synchronize', 'reopened']"
      bindings:
        - ref: github-pr-binding
        - name: gitrevision1
          value: $(body.pull_request.head.sha)
        - name: gitrepositoryurl1
          value: $(body.repository.clone_url)
```
and EL creation is success :thinking: 

But as per the validation it should not 
correct me if my understanding is wrong

File: docs/triggers.md
Comment (savitaashture) [inline docs/triggers.md:28]: is bindings are optional ?

File: pkg/apis/triggers/v1alpha1/trigger_validation.go
Comment (savitaashture) [inline pkg/apis/triggers/v1alpha1/trigger_validation.go:81]: When these conditions are applicable :thinking:

File: examples/gitlab/gitlab-push-listener.yaml
Comment (savitaashture) [inline examples/gitlab/gitlab-push-listener.yaml:55]: I see both `ref` and `name, value` are allowed together for binding, So can we update example to show such scenario?
something like below
```
      bindings:
        - ref: gitlab-push-binding
        - name: gitrevision
          value: $(body.pull_request.head.sha)
```

```
apiVersion: triggers.tekton.dev/v1alpha1
kind: TriggerBinding
metadata:
  name: gitlab-push-binding
spec:
  params:
    - name: gitrepositoryurl
      value: $(body.repository.git_http_url)
```

Comment (dibyom) [reply examples/gitlab/gitlab-push-listener.yaml:55]: Sure, I'm planning on changing all of our examples to use embedded bindings by default in a PR once this is merged. I'd prefer to create a new example to show that refs and embeds can be mixed and matched instead of adding it to this one example. What do you think of doing that in the followup PR?

File: pkg/apis/triggers/v1alpha1/trigger_validation.go
Comment (dibyom) [reply pkg/apis/triggers/v1alpha1/trigger_validation.go]: if we use Ref, we cannot also provide a Name. If we use Ref, we can optionally provide a Kind whose valid values are TriggerBinding or ClusterTriggerBinding. Does that answer your question? Or were you looking for something else?

File: examples/gitlab/gitlab-push-listener.yaml
Comment (savitaashture) [reply examples/gitlab/gitlab-push-listener.yaml:55]: Sure :+1:

**Review by wlynch — COMMENTED** [review]
/lgtm

Looks good, just some minor doc updates we should make to clarify usage from godocs.

File: pkg/apis/triggers/v1alpha1/trigger_types.go
Comment (dibyom) [reply pkg/apis/triggers/v1alpha1/trigger_types.go:55]: yeah its https://github.com/tektoncd/triggers/blob/c8123cf6d65cc87354b07a5376611f642eca3e5b/pkg/apis/triggers/v1alpha1/event_listener_types.go#L169

File: pkg/apis/triggers/v1alpha1/trigger_validation.go
Comment (dibyom) [reply pkg/apis/triggers/v1alpha1/trigger_validation.go:71]: good idea!

File: pkg/apis/triggers/v1alpha1/trigger_types.go
Comment (wlynch) [reply pkg/apis/triggers/v1alpha1/trigger_types.go:59]: ... I can't read 😅  

lgtm!

**Review by wlynch — APPROVED** [review]
/lgtm

**Pull request merged**
